enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+struct GenericOpInterfaceReverse
+    : public ReverseAutoDiffOpInterface::ExternalModel<
+          GenericOpInterfaceReverse, linalg::GenericOp> {
+  void createReverseModeAdjoint(Operation *op, OpBuilder &builder,
+                                MGradientUtilsReverse *gutils,
+                                SmallVector<Value> caches) const {
+    // TODO: currently in progress, this doesn't work
+    auto linalgOp = cast<linalg::LinalgOp>(op);
+    assert(linalgOp.hasBufferSemantics() &&
+           "Linalg op with tensor semantics not yet supported");
+    linalg::LinalgOp newOp =
+        cast<linalg::LinalgOp>(gutils->getNewFromOriginal(linalgOp));
+
+    auto cacheBuilder = OpBuilder(newOp);
+
+    // get iteration domain
+    AffineMap aMap = newOp.getShapesToLoopsMap();
+    SmallVector<Value> dims;
+    for (OpOperand *input : newOp.getInputOperands()) {
+      auto shape = cast<MemRefType>(input->get().getType()).getShape();
+      for (int i = 0; i < shape.size(); i++) {
+        auto dimI =
+            cacheBuilder.create<arith::ConstantIndexOp>(op->getLoc(), i);
+        // input->get().getParentBlock()->dump();
+        auto dim = cacheBuilder.create<memref::DimOp>(op->getLoc(),
+                                                      input->get(), dimI);
+        dims.push_back(dim);
+      }
+    }
+
+    SmallVector<Value> iterationDomains;
+    std::vector<int64_t> shapes;
+    for (unsigned int i = 0; i < aMap.getNumResults(); i++) {
+      AffineMap subMap = aMap.getSubMap({i});
+      Value domain = cacheBuilder.create<AffineApplyOp>(op->getLoc(), subMap,
+                                                        ValueRange(dims));
+      iterationDomains.push_back(domain);
+      shapes.push_back(-1);
+    }
+
+    SmallVector<Value> inputs, outputs;
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<StringRef> iteratorTypes{linalgOp.getNumLoops(),
+                                         getParallelIteratorTypeName()};
+
+    for (OpOperand *output : linalgOp.getOutputOperands()) {
+      if (gutils->hasInvertPointer(output->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(output));
+        Value out = gutils->invertPointerM(output->get(), builder);
+        Value view = invertMemref(out, builder, op->getLoc());
+        outputs.push_back(view);
+      }
+    }
+
+    for (OpOperand *input : linalgOp.getInputOperands()) {
+      if (gutils->hasInvertPointer(input->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(input));
+        Value inp = gutils->invertPointerM(input->get(), builder);
+        Value view = invertMemref(inp, builder, op->getLoc());
+        inputs.push_back(view);
+      }
+    }
+
+    linalg::GenericOp adjoint = builder.create<linalg::GenericOp>(
+        op->getLoc(), outputs, inputs, indexingMaps, iteratorTypes);
+
+    int numInputs = inputs.size();
+    std::function<buildReturnFunction> buildFuncReturnOp =
```

Comment:
`std::function` constructor performs dynamic allocation, which is bad for performance. Use `llvm::function_ref` or keep lambdas as `auto` unless (1) they are recursive or (2) the function-typed variable will outlive the function itself.